File: vitagraph/backend/app/schemas/graph.py
"""Schemas for Knowledge Graph nodes, edges, analytics, and subgraph queries."""

from __future__ import annotations

from typing import Any
from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    id: str
    label: str
    type: str  # report | test | measurement | chunk | category
    color: str
    r: float = 12.0
    betweenness: float = 0.0
    community: int = 0
    active: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    source: str
    target: str
    relation: str = "CONNECTED_TO"


class GraphMetrics(BaseModel):
    total_nodes: int
    total_edges: int
    communities_count: int
    modularity: float
    density: float


class GraphResponse(BaseModel):
    nodes: list[dict[str, Any]]
    edges: list[dict[str, Any]]
    metrics: GraphMetrics
    active_concepts: list[str] = Field(default_factory=list)


class SubgraphRequest(BaseModel):
    user_id: str
    chunk_ids: list[str]

File: vitagraph/frontend/src/api/graph.ts
import { api } from "./client";

export interface GraphNode {
  id: string;
  label: string;
  type: "report" | "test" | "measurement" | "chunk" | "category" | string;
  color: string;
  r?: number;
  betweenness?: number;
  community?: number;
  active?: boolean;
  test_name?: string;
  category?: string;
  value?: number;
  unit?: string;
  flag?: string;
  date?: string;
  page?: number;
  chunk_id?: string;
  report_id?: string;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation: string;
}

export interface GraphMetrics {
  total_nodes: number;
  total_edges: number;
  communities_count: number;
  modularity: number;
  density: number;
}

export interface GraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
  metrics: GraphMetrics;
  active_concepts?: string[];
}

export const graphApi = {
  getGraph: (userId: string) => api.get<GraphResponse>(`/api/graph/${userId}`),
  getSubgraph: (userId: string, chunkIds: string[]) =>
    api.post<GraphResponse>("/api/graph/subgraph", { user_id: userId, chunk_ids: chunkIds }),
};
